### PR TITLE
[WIP] Magma backend [magma-dev]

### PR DESCRIPTION
**Pull request by stomov**
Start a magma backend and check how everything works...

### CODE REVIEW

**Comment by jedbrown** [comment]
How does one build MAGMA without a CUDA device? And how is it run on a CPU?

**Comment by stomov** [comment]
MAGMA has a defined a generic interface to its building blocks, and provides wrappers (in that interface) for CUDA GPU, OpenCL, and CPU (including Phi), similar to the backends here. What is committed though now is just a copy of the reference backend (so there is no magma yet) - just to put the structure in place and get familiar with the interface, and next will start putting the functional magma backend.

**Comment by jedbrown** [comment]
Okay, so does it make sense to merge this?  Or is this branch/PR currently WIP?

**Comment by stomov** [comment]
No need to merge now - I can cancel the request and pull when the backend is complete.

**Comment by jedbrown** [comment]
I renamed the PR to have `[WIP]`. We can use it as a placeholder for discussion because it tracks the commits in this branch (an issue doesn't do that).

**Comment by stomov** [comment]
Jed, I am still trying to figure out the structure and sequence of calling things, e.g., who calls the Register function? I see it is in both Occa and Ref but 'make test' seems to only call the one in Ref. Thanks.

**Comment by jedbrown** [comment]
If you have OCCA installed, then the OCCA version will also be compiled (and registered). The `__attribute__((constructor))` causes it to be called automatically (before `main`).

**Comment by stomov** [comment]
Sorry I am not familiar with this mechanism. I see Occa backend and Ref are called at the same way but still it works for Ref but not for OCCA, e.g., 'make test' gives
tomovmac:libCEED tomov$ make test
1..15
build/t00-init /cpu/self
ok 1 t00-init /cpu/self
ok 2 t00-init /cpu/self stdout
ok 3 t00-init /cpu/self stderr
ok 4 # SKIP t00-init /cpu/occa
...
So, why the ref Register is called but not the OCCA one, and by "install" what do you mean in terms of linking it to libCEED (the installation will be independent of libCEED, so libCEED will not know about OCCA, right, or wrong)?

**Comment by jedbrown** [comment]
You don't have OCCA available so that backend is not registered.  If you
install OCCA (either to ../occa or specify via OCCA_DIR) and rebuild
libCEED, then it will also be registered and then those cases will not
be skipped.  The OCCA backend is not compiled and linked into libCEED
when OCCA isn't installed.

**Comment by tzanio** [comment]
@stomov, is this ready for review?

**Comment by tzanio** [comment]
This is ready for merge: @camierjs, @jedbrown and @v-dobrev please review.

**Comment by v-dobrev** [comment]
@stomov, @haidarazzam, I'm trying to build this branch and I get an error in `backends/magma/ceed-magma_cuda.cu` - it wants to `#include "atomics.cuh"` which is a header I cannot find.

**Comment by v-dobrev** [comment]
I found `atomics.cuh` in the MAGMA source. So, I guess, this branch requires MAGMA in-source build. Trying to build MAGMA in-source with `make`.

**Comment by v-dobrev** [comment]
I could not build with CUDA 9.1, so I switched to CUDA 8.0. That worked but I still had to apply the patch for gcc-4.8 that I added to the MAGMA Spack package. With that, I was able to build libceed with magma - there are some warnings but `make` finishes successfully. :smiley:

File: backends/magma/ceed-magma.c [hunk 1]
```diff
@@ -0,0 +1,806 @@
...
+    CeedInt P = basis->P1d, Q = basis->Q1d;
+    if (tmode == CEED_TRANSPOSE) {
+      P = basis->Q1d; Q = basis->P1d;
+    }
+    CeedInt pre = ndof*CeedPowInt(P, dim-1), post = 1;
+    CeedScalar tmp[2][ndof*Q*CeedPowInt(P>Q?P:Q, dim-1)];
+    CeedDebug("\033[01m[CeedBasisApply_Magma] tmpsize = %d",
+              ndof*Q*CeedPowInt(P>Q?P:Q, dim-1));
+    for (CeedInt d=0; d<dim; d++) {
+      ierr = CeedTensorContract_Magma(basis->ceed, pre, P, post, Q, basis->interp1d,
```
Comment (jedbrown) [inline backends/magma/ceed-magma.c]: So this is called once per element, and internally does `magma_malloc` and other relatively heavy operations for that granularity (especially if targeting a GPU). I had assumed that the MAGMA implementation would want hoist this call up a level to batch over many elements.

Comment (stomov) [reply backends/magma/ceed-magma.c]: Yes, definitely, this is the plan, but left this part for next release. For now we wanted to understand the flow, and have the functionality coverage on how magma will get plugged in, and have a way to fuse kernels over an element and have device code looking like the CPU code.

**Comment by tzanio** [comment]
I can almost build if I copy the following files to `spack location -i magma`/include:

```
 magma/magmablas/atomics.cuh
 magma/control/magma_threadsetting.h
 magma/control/pthread_barrier.h
 magma/control/magma_internal.h
```

@stomov: Are these required for the magma backend? Should they be added to the MAGMA Spack package?

I do get an error at the end for `libceed.so` though:
```
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -c -o build/ceed-basis.o /home/kolev1/ceed/libCEED/ceed-basis.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -c -o build/ceed-fortran.o /home/kolev1/ceed/libCEED/ceed-fortran.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -c -o build/ceed-elemrestriction.o /home/kolev1/ceed/libCEED/ceed-elemrestriction.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -c -o build/ceed-operator.o /home/kolev1/ceed/libCEED/ceed-operator.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -c -o build/ceed-vec.o /home/kolev1/ceed/libCEED/ceed-vec.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -c -o build/ceed.o /home/kolev1/ceed/libCEED/ceed.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -c -o build/ceed-qfunction.o /home/kolev1/ceed/libCEED/ceed-qfunction.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -c -o build/backends/ref/ceed-ref.o /home/kolev1/ceed/libCEED/backends/ref/ceed-ref.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -I/home/kolev1/mfem/occa/include -c -o build/backends/occa/ceed-occa-basis.o /home/kolev1/ceed/libCEED/backends/occa/ceed-occa-basis.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -I/home/kolev1/mfem/occa/include -c -o build/backends/occa/ceed-occa-okl.o /home/kolev1/ceed/libCEED/backends/occa/ceed-occa-okl.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -I/home/kolev1/mfem/occa/include -c -o build/backends/occa/ceed-occa-operator.o /home/kolev1/ceed/libCEED/backends/occa/ceed-occa-operator.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -I/home/kolev1/mfem/occa/include -c -o build/backends/occa/ceed-occa-qfunction.o /home/kolev1/ceed/libCEED/backends/occa/ceed-occa-qfunction.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -I/home/kolev1/mfem/occa/include -c -o build/backends/occa/ceed-occa-restrict.o /home/kolev1/ceed/libCEED/backends/occa/ceed-occa-restrict.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -I/home/kolev1/mfem/occa/include -c -o build/backends/occa/ceed-occa-vector.o /home/kolev1/ceed/libCEED/backends/occa/ceed-occa-vector.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -I/home/kolev1/mfem/occa/include -c -o build/backends/occa/ceed-occa.o /home/kolev1/ceed/libCEED/backends/occa/ceed-occa.c
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -DADD_ -I/home/kolev1/ceed/spack/opt/spack/linux-rhel7-x86_64/gcc-4.9.2/magma-2.3.0-texpekqkyi4f4t2b2xmysixcwlqrpton/include -I/home/kolev1/local/cuda/include -c -o build/backends/magma/magma_dtensor_contract.o /home/kolev1/ceed/libCEED/backends/magma/magma_dtensor_contract.c
python backends/magma/gccm.py backends/magma/ceed-magma.c
backends/magma/gccm.py backends/magma/ceed-magma.c --> backends/magma/ceed-magma_tmp.c backends/magma/ceed-magma_cuda.cu
gcc -I./include -std=c99 -Wall -Wextra -Wno-unused-parameter -fPIC -MMD -MP -O2 -g -DNDEBUG=1 -DUNDERSCORE  -DADD_ -I/home/kolev1/ceed/spack/opt/spack/linux-rhel7-x86_64/gcc-4.9.2/magma-2.3.0-texpekqkyi4f4t2b2xmysixcwlqrpton/include -I/home/kolev1/local/cuda/include -c -o build/backends/magma/ceed-magma_tmp.o /home/kolev1/ceed/libCEED/backends/magma/ceed-magma_tmp.c
nvcc -I./include --compiler-options=-fPIC -DADD_ -I/home/kolev1/ceed/spack/opt/spack/linux-rhel7-x86_64/gcc-4.9.2/magma-2.3.0-texpekqkyi4f4t2b2xmysixcwlqrpton/include -I/home/kolev1/ceed/spack/opt/spack/linux-rhel7-x86_64/gcc-4.9.2/magma-2.3.0-texpekqkyi4f4t2b2xmysixcwlqrpton/magmablas -I/home/kolev1/ceed/spack/opt/spack/linux-rhel7-x86_64/gcc-4.9.2/magma-2.3.0-texpekqkyi4f4t2b2xmysixcwlqrpton/control -I/home/kolev1/local/cuda/include -c -o build/backends/magma/ceed-magma_cuda.o /home/kolev1/ceed/libCEED/backends/magma/ceed-magma_cuda.cu
gcc    -L/home/kolev1/mfem/occa/lib -Wl,-rpath,/home/kolev1/mfem/occa/lib -L/home/kolev1/ceed/spack/opt/spack/linux-rhel7-x86_64/gcc-4.9.2/magma-2.3.0-texpekqkyi4f4t2b2xmysixcwlqrpton/lib -Wl,-rpath,/home/kolev1/ceed/spack/opt/spack/linux-rhel7-x86_64/gcc-4.9.2/magma-2.3.0-texpekqkyi4f4t2b2xmysixcwlqrpton/lib -shared -o lib/libceed.so build/ceed-basis.o build/ceed-fortran.o build/ceed-elemrestriction.o build/ceed-operator.o build/ceed-vec.o build/ceed.o build/ceed-qfunction.o build/backends/ref/ceed-ref.o build/backends/occa/ceed-occa-basis.o build/backends/occa/ceed-occa-okl.o build/backends/occa/ceed-occa-operator.o build/backends/occa/ceed-occa-qfunction.o build/backends/occa/ceed-occa-restrict.o build/backends/occa/ceed-occa-vector.o build/backends/occa/ceed-occa.o build/backends/magma/magma_dtensor_contract.o build/backends/magma/ceed-magma_tmp.o build/backends/magma/ceed-magma_cuda.o -lm -locca -lmagma -L/home/kolev1/local/cuda/lib64 -lcudart
/usr/bin/ld: /home/kolev1/ceed/spack/opt/spack/linux-rhel7-x86_64/gcc-4.9.2/magma-2.3.0-texpekqkyi4f4t2b2xmysixcwlqrpton/lib/libmagma.a(copy_v2.cpp.o): relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
/home/kolev1/ceed/spack/opt/spack/linux-rhel7-x86_64/gcc-4.9.2/magma-2.3.0-texpekqkyi4f4t2b2xmysixcwlqrpton/lib/libmagma.a: error adding symbols: Bad value
collect2: error: ld returned 1 exit status
make: *** [lib/libceed.so] Error 1
```
Any suggestions how to fix?

**Comment by tzanio** [comment]
The above errors go away if magma is build shared. I can build everything now.

**Comment by tzanio** [comment]
I will make the changes in the MAGMA Spack package... no other changes required here for me

File: backends/magma/ceed-magma.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+#define USE_MAGMA_BATCH
+
+void magma_dtensor_contract(Ceed ceed,
+                            CeedInt A, CeedInt B, CeedInt C, CeedInt J,
+                            const CeedScalar *t, CeedTransposeMode tmode,
+                            const CeedInt Add,
+                            const CeedScalar *u, CeedScalar *v);
+
+#define CeedDebug(...)
```
[the remark this answers is not in the document]
Comment (camierjs) [reply backends/magma/ceed-magma.h]: We can just leave it like that for now, we'll unify the CeedDebug function later

**Comment by tzanio** [comment]
I can't reproduce @camierjs's problem: on my machine both `-c /gpu/occa` and `-c /gpu/magma` work. The latter doesn't give correct results, but it works.

**Comment by stomov** [comment]
@tzanio  @camierjs  I would also have suggested to just copy atomics.cuh in backends/magma for now, as @camierjs did. For the future we will put these include files at the right places in magma.

**Comment by tzanio** [comment]
@stomov -- in general we can't copy  `atomics.cuh`, `magma_internal.h`, `pthread_barrier.h` and `magma_threadsettings.h` to `backends/magma` because we will want this to work with the Spack install (which removes the  MAGMA source after installation).

Are you OK with copying them to the `magma/install` directory as in https://github.com/spack/spack/pull/7423/commits/18e74addb2e2aefd9234d2e5987d082d9fefd639?

**Comment by stomov** [comment]
@tzanio yes, this sounds good.

**Comment by jedbrown** [comment]
It looks like `CeedVectorGetArray(mesh_coords, CEED_MEM_HOST, &coords);` is not correct, even with `/cpu/magma`.

```
~/src/libceed/examples/ceed$ valgrind ./ex1 -ceed /cpu/magma
==32446== Memcheck, a memory error detector
==32446== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==32446== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==32446== Command: ./ex1 -ceed /cpu/magma
==32446==
==32446== Warning: ignored attempt to set SIGKILL handler in sigaction();
==32446==          the SIGKILL signal is uncatchable
Selected options: [command line option] : <current value>
  Ceed specification [-c] : /cpu/magma
  Mesh dimension     [-d] : 3
  Mesh order         [-m] : 4
  Solution order     [-o] : 4
  Num. 1D quadr. pts [-q] : 6
  Approx. # unknowns [-s] : 262144

Mesh size: nx = 16, ny = 16, nz = 16
Number of mesh nodes    : 274625
Number of solution dofs : 274625
==32446== Use of uninitialised value of size 8
==32446==    at 0x10A8D6: SetCartesianMeshCoords (ex1.c:365)
==32446==    by 0x109F25: main (ex1.c:188)
==32446==
==32446== Invalid write of size 8
==32446==    at 0x10A8D6: SetCartesianMeshCoords (ex1.c:365)
==32446==    by 0x109F25: main (ex1.c:188)
==32446==  Address 0x1fff217f08 is not stack'd, malloc'd or (recently) free'd
```

**Comment by tzanio** [comment]
@jedbrown: do you see any build issues with magma and occa enabled in libCEED like @camierjs did in https://github.com/CEED/libCEED/pull/36#issuecomment-377427293?

**Comment by camierjs** [comment]
Ok, I was using another CEED_DIR in the example/ceed directory.
So the output I was seeing was the result of the CeedInit "matchidx" logic: ex1 -c /gpu/* now triggers the occa backend.

**Comment by jedbrown** [comment]
@v-dobrev Does it pass all the automated tests for you? I get SEGV in tests too, with `/cpu/magma` backend.

**Comment by v-dobrev** [comment]
When I tested this PR a few days ago (using an in-source build of magma, not the one build through spack) I was getting errors in `make test` only in the libceed example - the value of the printed volume was some big number (something like `...e+190`).

**Comment by tzanio** [comment]
I agree with @v-dobrev and am OK with his plan, unless @stomov really wants to fix the results before the merge.

@jedbrown: `make prove` on my machine fails for `t30-operator` and `ex1`, tests 17, 20.

**Comment by jedbrown** [comment]
@stomov I get SEGV in the tests.

```
$ gdb -ex r --args build/t30-operator /cpu/magma                                                                                                                
No symbol table is loaded.  Use the "file" command.                                                                                                                                           
Breakpoint 1 (PetscError) pending.
Reading symbols from build/t30-operator...done.                                                                                                                                               
Starting program: /home/jed/src/libceed/build/t30-operator /cpu/magma                                                                                                                         
[Thread debugging using libthread_db enabled]                                                                                                                                                 
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[New Thread 0x7fffea741700 (LWP 9516)]                                                                                                                                                        
                                                                                                                                                                                              
Thread 1 "t30-operator" received signal SIGSEGV, Segmentation fault.                                                                                                                          
0x0000000000000000 in ?? ()                                                                                                                                                                   
(gdb) bt                                                                                                                                                                                      
#0  0x0000000000000000 in ?? ()                                                                                                                                                               
#1  0x00007ffff7bb894f in CeedElemRestrictionApply (r=0x5555558b6660, tmode=CEED_NOTRANSPOSE, ncomp=1, lmode=CEED_NOTRANSPOSE, u=0x5555558b7000, v=0x5555558b73f0, request=0x7ffff7dcf480 <ceed_request_immediate>) at /home/jed/src/libceed/ceed-elemrestriction.c:119
#2  0x00007ffff7bc5912 in CeedOperatorApply_Magma (op=0x5555558b7690, qdata=0x5555558b6fb0, ustate=0x5555558b7000, residual=0x0, request=0x7ffff7dcf480 <ceed_request_immediate>) at /home/jed/src/libceed/backends/magma/ceed-magma_tmp.c:718
#3  0x00007ffff7bb8bd6 in CeedOperatorApply (op=0x5555558b7690, qdata=0x5555558b6fb0, invec=0x5555558b7000, outvec=0x0, request=0x7ffff7dcf480 <ceed_request_immediate>) at /home/jed/src/libceed/ceed-operator.c:82
#4  0x0000555555555493 in main (argc=2, argv=0x7fffffffda08) at /home/jed/src/libceed/tests/t30-operator.c:69                                                                                 
(gdb) up                                                                                                                                                                                      
#1  0x00007ffff7bb894f in CeedElemRestrictionApply (r=0x5555558b6660, tmode=CEED_NOTRANSPOSE, ncomp=1, lmode=CEED_NOTRANSPOSE, u=0x5555558b7000, v=0x5555558b73f0, request=0x7ffff7dcf480 <ceed_request_immediate>) at /home/jed/src/libceed/ceed-elemrestriction.c:119
119       ierr = r->Apply(r, tmode, ncomp, lmode, u, v, request); CeedChk(ierr);                                                                                                              
(gdb) p r                                                                                                                                                                                     
$1 = (CeedElemRestriction) 0x5555558b6660                                                                                                                                                     
(gdb) p *r                                                                                                                                                                                    
$2 = {                                                                                                                                                                                        
  ceed = 0x5555558b65d0,                                                                                                                                                                      
  Apply = 0x0,                                                                                                                                                                                
  Destroy = 0x0,                                                                                                                                                                              
  refcount = 2,                                                                                                                                                                               
  nelem = 5, 
  elemsize = 2, 
  ndof = 6, 
  data = 0x0
}
```

**Comment by jedbrown** [comment]
@stomov The issue here is that `magma_malloc` can fail. I'm guessing that's because it's trying to get device memory even though I asked for `/cpu/magma`.

**Comment by v-dobrev** [comment]
Maybe we can remove (temporarily?) `/cpu/magma` - leave just `/gpu/magma`?

Or, we can just document that both are experimental / WIP.

**Comment by jedbrown** [comment]
Discussion is disjoint between here and the team thread, but would you be opposed to tagging v0.2 and merging the magma backend in its current state immediately after tagging. Anyone who wants to use the magma backend would need to use 'master' anyway (because it segfaults or gives incorrect results even within our test suite). That way we can feel good about the release and still say in the report that experimental MAGMA support has been merged. Since the presence of MAGMA does not affect the public interface, we could even provide it in v0.2.1 when the bugs are ironed out.

**Comment by haidarazzam** [comment]
the magma pass the tests on our laptop
it might be something minor.

for magma_malloc thats the GPU interface we should have a magma_malloc _cpu for the CPU one
Stan and I will be working on it now to try to fix it.
Azzam

**Comment by stomov** [comment]
Actually, it was passing the tests before but after I pulled with the latest changes in backends/ref/ceed-ref.c  and tests/t30-operator.c we get errors, even if we disregard of use of MAGMA code. So there must have been a bug before in the reference that got fixed and we didn't propagate it in the magma backend or something like this. I am also going back now and comparing to see what changed.

**Comment by jedbrown** [comment]
@stomov On current 'master', the only change in `backends/ref` since 8c081a34cb1041aee9718298ff78f7dc1fc6b5dc (January 8) is a copyright notice.

**Comment by jedbrown** [comment]
@stomov Same in your branch -- no changes in backends/ref since early January. @camierjs had some changes and un-changes in his branch, but eventually ended up with a no-op.

**Comment by v-dobrev** [comment]
The question is what are the changes in master since @stomov branched out `magma-dev` and if `master` has been merged into `magma-dev` after that breaking what was a working (for him) branch.

**Comment by v-dobrev** [comment]
I was testing 5a9ca9a when I saw only two failures in `make test` - both in what was then `ex1-volume-annulus`.

**Comment by jedbrown** [comment]
Yes, that's why merging from upstream should be done cautiously and only when truly necessary. My point is that any such changes are not in backends/ref. Unless there are objections in the next few minutes, I'm tagging v0.2 without this branch and we can merge it just afterward (in its current state or after bug fixes).

**Comment by jedbrown** [comment]
@v-dobrev The relevant change since then is probably adding these lines.

```
  CeedVectorGetArrayRead(V, CEED_MEM_HOST, &hv);
  for (CeedInt i=0; i<Nu; i++)
    if (hv[i] != 0.0) printf("[%d] v %f != 0.0\n",i, hv[i]);
```

(i.e., actually checking the result)

**Comment by v-dobrev** [comment]
OK, let's give @stomov and @haidarazzam some time - I think they are looking into this right now.

**Comment by tzanio** [comment]
@stomov and @haidarazzam -- are you still looking at this? Should we give you a couple of more hours, or should we go ahead with releasing 0.2?

**Comment by haidarazzam** [comment]
I am trying to go through the commit and find out what did cause this.
if this is just a tag on github, then can’t we do another tag once done or sometimes next week when we fix all the issues.


Azzam

**Comment by jedbrown** [comment]
@haidarazzam Yes, that's definitely fine. v0.2 is now tagged. Thanks, everyone.

**Comment by stomov** [comment]
I have a question on the API. When a vector is created, we want to be able to use either CPU, GPU or combination, so we keep two pointers - one to CPU memory and one to GPU memory. In order to do this we need to implement some coherency protocol - we can give pointer to the data to CeedVectorGetArray either on CPU or DEVICE, but have to know if that data was modified and when. I thought this is the idea of CeedVectorRestoreArray in the API, but this is not consistently called, e.g., in the CeedOperatorApply qdata gives pointer to qd, qd gets modified but  CeedVectorRestoreArray is not called. There are a few other cases that this is happening. If I "fix" them I pass the tests but I am not sure if my interpretation of the intent of the API is correct. Any comment on this?

**Comment by jedbrown** [comment]
That missing Restore is indeed a (cosmetic at present, given its usage) bug in the reference backend. Every Get should be paired with a Restore, I just haven't implemented the debug-mode checking of that.

The philosophy was that the backend knows which device it wants to use because that is in the resource spec passed to `CeedInit_Xyz`. The caller can provide memory on the host or device and can ask for it on the host or device, so the backend needs to be able to provide that on demand. It's up to the backend whether to persist that representation after the Restore.

**Comment by stomov** [comment]
We managed to debug the magma backend and now we pass all tests and ex1. I pushed the changes. The issues were:
1) some missing VectorRestores (now added);
2) t30 was assuming vector allocation initializes it with 0s (wasn't true for GPU allocation);
3) bug fixes from Azzam (thanks!).
@jedbrown @v-dobrev @tzanio  thanks for answering the questions above and the comments!

**Comment by tzanio** [comment]
Can we merge this as an initial version, and work on fixing/improving the performance in the next release?

I'd like to tag `v0.2.1` so "initial MAGMA support" is present in the `v0.2` series.

**Comment by v-dobrev** [comment]
I'm fine with merging this - it was working correctly when I last tested.

**Comment by jedbrown** [comment]
Still failing with `/cpu/magma` for me.
```
not ok 13 t11-basis /cpu/magma 
ok 14 t11-basis /cpu/magma stdout
not ok 15 t11-basis /cpu/magma stderr
# +/home/jed/src/libceed/backends/magma/ceed-magma_tmp.c:829 in CeedInit_Magma(): error in magma_init(): -116
# +                                  
1..15        
```

How should I go about finding more detailed diagnostics?

**Comment by stomov** [comment]
/cpu/magma and /gpu/magma are the same, so they should work the same way if there is GPU with CUDA on the system.  The particular error is in magma_init and error code is for MAGMA_ERR_UNKNOWN or unspecified error.

**Comment by jedbrown** [comment]
How can I help you debug why `/cpu/magma` is crashing on my computer?

**Comment by v-dobrev** [comment]
@jedbrown, I was getting the same error when I was trying to use a newer version of CUDA which does not work with the installed nvidia driver. So my guess is that you need to either update your nvidia driver or use an older version of CUDA.

Also: as @stomov currently there is no difference which string you specify - both `/cpu/magma` and `/gpu/magma` use the same kernels, so you do need cuda even if you specify `/cpu/magma`.

**Comment by jedbrown** [comment]
If `/cpu/magma` actually runs on a GPU then it should be removed instead of lying to the user. If it doesn't run on a GPU, then it should be possible to run without a GPU. (I'm not concerned about whether CUDA is needed to compile at this point.)

**Comment by v-dobrev** [comment]
Yes `/cpu/magma` is confusing. @stomov, do you plan to make that specification do something different from `/gpu/magma`? If not, then I agree that it is better to remove it now.

**Comment by stomov** [comment]
I also agree and just made the change.

**Comment by camierjs** [comment]
I still have to drop the last three header files (magma_internal, pthread_barrier & magma_threadsetting) to compile when magma was compiled and installed manually (not from spack).
The TAP tests are fine but `examples/ceed` with /gpu/magma is still long (~40s) and outputs a null volume, mfem bp1 takes ~2mn without converging either.

**Comment by tzanio** [comment]
> I still have to drop the last three header files (magma_internal, pthread_barrier & magma_threadsetting) to compile when magma was compiled and installed manually (not from spack).
This should be addressed by the MAGMA build system, @stomov do you have plans for that?

> The TAP tests are fine but examples/ceed with /gpu/magma is still long (~40s) and outputs a null volume, mfem bp1 takes ~2mn without converging either.
I believe @v-dobrev reported correct results (though taking a long time)

**Comment by v-dobrev** [comment]
I checked again with the latest version of this branch and apart from being slow, I do not see any issues - including `examples/ceed/ex1`, `examples/mfem/bp1`, and `examples/nek5000/bp1`.

**Comment by v-dobrev** [comment]
@jedbrown, the `t30-operator.c` test was not initializing the vector `U`, so I added explicit initialization. Was this a bug, or did you expect `U` to be initialized with zeros by the backend when the actual memory allocation happens?

**Comment by jedbrown** [comment]
@v-dobrev Thanks for finding that. Yeah, explicit initialization is desirable, but we should add an interface, probably `CeedVectorSet(CeedVector, CeedScalar)`. I include the value because there are occasional circumstances where a vector of ones (or other scaling factor) is useful.

**Comment by v-dobrev** [comment]
Yes, I agree. The way I implemented it - with `CeedVectorGetArray` and `CeedVectorRestoreArray` - is just a quick fix.

**Comment by tzanio** [comment]
@jedbrown any objections to meging?

**Pull request merged**